### PR TITLE
Migrations: Finalize CUD APIs

### LINKED ISSUE #8471: Review and update CUD APIs for migrations
As part of the data seeding work now tracked in https://github.com/aspnet/EntityFramework/issues/629, we created CUD APIs that can be used inside migrations. 

The data seeding work is currently not scheduled to make it in 2.0, but the CUD APIs are already in.

Before we release 2.0.0 we should make sure the CUD APIs are in the state we want them to be, e.g.

1. Are compatible with the usage in the current PR https://github.com/aspnet/EntityFramework/pull/7824
2. Are very usable and we are confident we won't want to make breaking changes on them after we ship.

Otherwise we should consider either fixing them or pulling them from the release.

cc @ajcvickers

### PR DESCRIPTION
**Pull request by bricelam**
Fixes #8471

### CODE REVIEW

File: src/EFCore.Relational/Migrations/MigrationBuilder.cs [hunk 1]
```diff
@@ -682,11 +682,20 @@ public virtual AlterOperationBuilder<AlterDatabaseOperation> AlterDatabase()
 
         public virtual OperationBuilder<InsertOperation> Insert(
```
Comment (bricelam) [inline src/EFCore.Relational/Migrations/MigrationBuilder.cs]: Will rename to `InsertData`, etc.

File: src/EFCore.Relational/Migrations/MigrationBuilder.cs [hunk 2]
```diff
@@ -697,18 +706,27 @@ public virtual AlterOperationBuilder<AlterDatabaseOperation> AlterDatabase()
...
-            [CanBeNull] object[,] keyValues = null)
+            [NotNull] string[] keyColumns,
+            [NotNull] object[] keyValues,
+            [CanBeNull] string schema = null)
+            => DeleteData(table, keyColumns, ToMultidimensionalArray(Check.NotNull(keyValues, nameof(keyValues))), schema);
+
+        public virtual OperationBuilder<DeleteDataOperation> DeleteData(
+            [NotNull] string table,
+            [NotNull] string[] keyColumns,
+            [NotNull] object[,] keyValues,
```
Comment (bricelam) [inline src/EFCore.Relational/Migrations/MigrationBuilder.cs]: I wonder if we should have overloads for a single key column. We do this in other methods that take key columns.

Comment (ajcvickers) [inline src/EFCore.Relational/Migrations/MigrationBuilder.cs]: Sounds reasonable. @divega?

Comment (divega) [inline src/EFCore.Relational/Migrations/MigrationBuilder.cs]: If it is sugar for the common case, sounds great.

Comment (bricelam) [inline src/EFCore.Relational/Migrations/MigrationBuilder.cs]: Updated. Love the terseness!

**Comment by bricelam** [comment]
PR submitted: dotnet/templating#851

**Comment by bricelam** [comment]
@divega Here's how I see it:

Now
===
The base `MigrationsSqlGenerator` will use the `UpdateDataOperation` objects directly to generate standard DML. If necessary (unlikely), providers can override the methods in their Migrations SQL generator to tweak it.

Later
===
We update our implementation in `MigrationsSqlGenerator` to convert the `UpdateDataOperation` objects into `ModificationCommandBase` objects and use the Update pipeline to generate the DML. Providers can safely remove their Migrations overrides since their Update pipeline implementations will become the source of truth.

**Comment by divega** [comment]
@bricelam I haven't looked at https://github.com/aspnet/EntityFramework/pull/7824/ in too much detail, but maybe you have :smile: 

Did you see anything specific there that makes you think we shouldn't make the change to `IUpdateSqlGenerator` in 2.0 or do you just think it is worth doing it?

**Comment by bricelam** [comment]
@AndriySvyryd Would know better about how much work is required, but I definitely think it's worth doing if we have time. The changes to Migrations are trivial (just converting from one data structure to another).

**Pull request merged**